[Spring ecommerce Order Step 2] Submit Bojana's mission

bojana-petroska opened the pull request:
**Implement Order Placement and Payment Integration with Stripe**

**Description:**

This PR introduces the core functionality for placing orders and processing payments using Stripe, along with necessary refactors, configurations, and deployment preparations.

Features Added:

**Order Module**
- Implemented `Order` and `OrderItem` entities.
- API to place and retrieve user orders.
- Order details include: order date/time, status, checkout session ID, items, and payment amount.

**Stripe Integration**
- Integrated `Stripe Payment API` for creating checkout sessions.
- Implemented `StripeClient`, config properties, and repository for payment tracking.
- Safe error handling for payment failures with descriptive feedback.

**Cart Updates**
- Automatically removes purchased items from the user's cart.
- Decreases stock of purchased product options accordingly.

**Refactoring and Improvements**
- Refactored `DTOs` and package names across cart, product, option, and member modules.
- Modified `addOptions` method to replace init method for product options.
- Improved error handling with enhanced `ErrorResponse` structure.
- Cleaned up optional fields in the product module.

**Deployment**
- Deployment script created for `AWS`.

san-ghun committed:
init: initiate project from ecommerce-product_api

san-ghun committed:
feat(model,repository): alter models into entities for JPA and apply 'ProductRepository'

san-ghun committed:
refactor: alter dao into repository by applying jpa

san-ghun committed:
style(model): reposition for 'id' property into last order

san-ghun committed:
refactor(Member,Cart): reformation the relation between 'Member' and 'Cart'

san-ghun committed:
refactor(CartStatisticsService): introduce 'CartStatisticsService'

san-ghun committed:
feat(Product): introduce pagination for products

san-ghun committed:
chore(sql): refine sql scripts

san-ghun committed:
test: refine test with JPA using TestEntityManager and LocalServerPort annotation

san-ghun committed:
feat(Product,CartItem): add pagination to products and cart items

san-ghun committed:
feat(Option): implement new entity 'Option' for 'Product'

- lots of test refactoring (sorry)

san-ghun committed:
docs(readme): update features list

bojana-petroska committed:
refactor(error): standardize API Error Responses, adjust the test cases

bojana-petroska committed:
refactor(cart): add sortBy with nested properties from Product

bojana-petroska committed:
refactor(cart, product, member): remove dependencies on DTOs in domain objects, move business logic into service, adjust tests

bojana-petroska committed:
refactor(product, option): change the logic for update and nameExists in ProductService, add Test for Product & Option

bojana-petroska committed:
refactor(product, option): remove unnecessary saving to the option repository, adjust the database to have options, remove the method addOptions

bojana-petroska committed:
refactor(view): update the fetched products to be displayed correctly

bojana-petroska committed:
refactor(product): remove unnecessary method

bojana-petroska committed:
refactor(cart, member, product): rename of the package for DTOs

bojana-petroska committed:
refactor(app, data): add dependencies for stripe, update readme, add error handling

bojana-petroska committed:
refactor(product): remove optional

bojana-petroska committed:
refactor(cart, data, product, option): change init into addOptions, fix failing tests, add changed package names

bojana-petroska committed:
feat(payment): add Stripe Client, Properties and Repository

bojana-petroska committed:
refactor(config, payment): add serialization, adjust ErrorResponse

bojana-petroska committed:
feat(order): create Order and Order Item module

bojana-petroska committed:
refactor(order): solve merge conflicts

Wordbe commented on build.gradle.kts:
```diff
@@ -2,6 +2,7 @@ plugins {
     kotlin("jvm") version "1.9.25"
     kotlin("plugin.spring") version "1.9.25"
     kotlin("plugin.jpa") version "1.9.25"
+    kotlin("plugin.serialization") version "1.8.0"
```
Why is this plugin needed? 🙂

bojana-petroska replied:
I used `serialization` to parse JSON error responses from Stripe into data classes like `StripeErrorResponse` and `StripeErrorMessage`

Wordbe commented on build.gradle.kts:
```diff
@@ -41,6 +41,8 @@ dependencies {
 
     testRuntimeOnly("org.junit.platform:junit-platform-launcher")
+
+    implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:1.5.1")
```
Here as well.

bojana-petroska replied:
This dependency complements the plugin and is used to parse JSON responses from Stripe

Wordbe commented on src/main/kotlin/ecommerce/GlobalExceptionHandler.kt:
```diff
@@ -70,4 +73,36 @@ class GlobalExceptionHandler {
     }
+
+    @ExceptionHandler(ProductNameAlreadyExistsException::class)
+    fun handleProductNameAlreadyExistsExceptionHandler(e: ProductNameAlreadyExistsException): ResponseEntity<ErrorResponse> {
+        val error = ErrorMessage("name", e.message ?: "Product name already exists..")
+        val errorResponse = ErrorResponse(listOf(error))
+        println("ProductNameAlreadyExistsException occurred: $errorResponse")
+        return ResponseEntity.badRequest().body(errorResponse)
+    }
+
+    @ExceptionHandler(MemberEmailAlreadyExistsException::class)
+    fun handleMemberEmailAlreadyExistsExceptionHandler(e: MemberEmailAlreadyExistsException): ResponseEntity<ErrorResponse> {
+        val error = ErrorMessage("name", e.message ?: "Member email already exists..")
+        val errorResponse = ErrorResponse(listOf(error))
+        println("MemberEmailAlreadyExistsException occurred: $errorResponse")
+        return ResponseEntity.badRequest().body(errorResponse)
+    }
+
+    @ExceptionHandler(StripeClientException::class)
+    fun handleStripeClientException(e: StripeClientException): ResponseEntity<ErrorResponse> {
+        println("StripeClientException occurred:" + e.message)
+        val stripeError = e.stripeErrorResponse.error
+        val errorResponse =
+            ErrorResponse(
+                listOf(
+                    ErrorMessage(
+                        stripeError.code ?: "unknown_code",
+                        stripeError.message,
+                    ),
+                ),
+            )
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse)
```
What about setting the error message as a default value in `ProductNameAlreadyExistsException` and `MemberEmailAlreadyExistsException`? Also, try using logging instead of `println`, it would be good to explore the differences and see which is better.

bojana-petroska replied:
I agree! Setting default messages directly in the exception constructors makes the handler code cleaner and avoids repetitive fallback logic. Also, regarding the logger `warn/error/debug` I already did it more deeply in our group project and will try to apply it here! : )

Wordbe commented on src/main/kotlin/ecommerce/dto/auth/AuthResponse.kt:
```diff
@@ -0,0 +1,5 @@
+package ecommerce.dto.auth
+
+data class AuthResponse(
```
Why did you use a `data class`? 🙂

bojana-petroska replied:
I used a data class here because `AuthResponse` is a simple `DTO` only to carry response data between layers

Wordbe replied:
But I think it might be fine to just use a regular class, what do you think? 🙂

Wordbe commented on src/main/kotlin/ecommerce/dto/cart/CartAddItemForm.kt:
```diff
@@ -0,0 +1,10 @@
+
+import jakarta.validation.constraints.Min
+
+data class CartAddItemForm(
```
Here as well.

bojana-petroska replied:
I used data class for all `DTO`s because they are simple, immutable data carriers between layers, and data class is idiomatic in Kotlin for this purpose.
Since the `DTO`s don’t contain any behavior or internal state and only hold data, I chose to use them.

Wordbe commented on src/main/kotlin/ecommerce/stripe/StripeClient.kt:
```diff
@@ -0,0 +1,88 @@
+        }
+    }
+
+    private fun parseStripeError(jsonString: String?): StripeErrorResponse? {
+        if (jsonString.isNullOrBlank()) return null
+        return try {
+            json.decodeFromString<StripeErrorResponse>(jsonString)
+        } catch (ex: Exception) {
+            null
+        }
+    }
```
How about standardizing JSON → object conversion with Jackson? 🙂

bojana-petroska replied:
That's a valid point!
I used kotlinx.serialization here primarily because I wanted direct JSON parsing for Stripe error responses.
On the other hand since the rest of the application uses Jackson, unifying on a single JSON library would make it more consistent, so I decided to change it and apply Jackson.

Wordbe commented on src/main/kotlin/ecommerce/stripe/StripeClient.kt:
```diff
@@ -0,0 +1,88 @@
+            "processing_error" -> "Payment failed due to a processing error."
+            else -> error?.message ?: "Payment failed for an unknown reason."
+        }
+    }
```
How about defining it as an enum? 🙂

Wordbe commented on src/main/kotlin/ecommerce/dto/cart/CartItemResponse.kt:
```diff
@@ -0,0 +1,10 @@
+import ecommerce.dto.product.ProductResponse
+import java.time.LocalDateTime
+
+data class CartItemResponse(
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/dto/cart/CartUpdateQuantityForm.kt:
```diff
@@ -0,0 +1,8 @@
+
+import jakarta.validation.constraints.Min
+
+data class CartUpdateQuantityForm(
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/dto/errors/StripeErrorResponse.kt:
```diff
@@ -0,0 +1,16 @@
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class StripeErrorResponse(
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/dto/errors/StripeErrorResponse.kt:
```diff
@@ -0,0 +1,16 @@
+)
+
+@Serializable
+data class StripeErrorMessage(
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/dto/member/ActiveMemberInfo.kt:
```diff
@@ -0,0 +1,7 @@
+package ecommerce.dto.member
+
+data class ActiveMemberInfo(
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/dto/member/LoginForm.kt:
```diff
@@ -0,0 +1,16 @@
+
+import jakarta.validation.constraints.NotBlank
+
+data class LoginForm(
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/dto/member/MemberResponse.kt:
```diff
@@ -0,0 +1,6 @@
+package ecommerce.dto.member
+
+data class MemberResponse(
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/dto/member/RegisterForm.kt:
```diff
@@ -0,0 +1,13 @@
+import jakarta.validation.constraints.NotBlank
+import jakarta.validation.constraints.Size
+
+data class RegisterForm(
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/dto/order/OrderRequest.kt:
```diff
@@ -0,0 +1,9 @@
+package ecommerce.dto.order
+
+data class OrderRequest(
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/dto/order/OrderRequest.kt:
```diff
@@ -0,0 +1,9 @@
+    val optionId: Long,
+    val quantity: Int,
+    val paymentMethod: String,
+    val currency: String = "usd",
```
What about trying to use an enum? 🟢

Wordbe commented on src/main/kotlin/ecommerce/dto/order/OrderResponse.kt:
```diff
@@ -0,0 +1,7 @@
+package ecommerce.dto.order
+
+data class OrderResponse(
```
.

Wordbe commented on src/main/kotlin/ecommerce/dto/payment/PaymentRequest.kt:
```diff
@@ -0,0 +1,7 @@
+package ecommerce.dto.payment
+
+data class PaymentRequest(
```
.

Wordbe commented on src/main/kotlin/ecommerce/dto/payment/PaymentResponse.kt:
```diff
@@ -0,0 +1,12 @@
+
+import com.fasterxml.jackson.annotation.JsonProperty
+
+data class PaymentResponse(
```
.

Wordbe commented on src/main/kotlin/ecommerce/dto/product/OptionResponse.kt:
```diff
@@ -0,0 +1,7 @@
+package ecommerce.dto.product
+
+data class OptionResponse(
```
.

Wordbe commented on src/main/kotlin/ecommerce/dto/product/ProductForm.kt:
```diff
@@ -0,0 +1,29 @@
+import jakarta.validation.constraints.Positive
+import jakarta.validation.constraints.Size
+
+data class ProductForm(
```
.

Wordbe commented on src/main/kotlin/ecommerce/dto/product/ProductForm.kt:
```diff
@@ -0,0 +1,29 @@
+    @field:NotBlank(message = "Product name is required")
+    @field:Pattern(regexp = "^[a-zA-Z0-9()\\[\\]+\\-&/_ ]+$", message = "Contains unallowed character")
+    @field:Size(min = 1, max = 15, message = "Must be no more than 15 characters, including spaces")
+    var name: String = String(),
+    @field:Positive(message = "Product price must be greater than zero")
+    var price: Double = Double.NaN,
+    @field:NotBlank(message = "Product image URL is required")
+    @field:Pattern(regexp = "^https?://.*", message = "Must start with 'https://'.")
+    var imageUrl: String = String(),
```
For an empty string, `""` seems more intuitive than `String()`.

Wordbe commented on src/main/kotlin/ecommerce/dto/product/ProductForm.kt:
```diff
@@ -0,0 +1,29 @@
+    var imageUrl: String = String(),
+) {
+    companion object {
+        fun toProduct(
+            productForm: ProductForm,
+            options: List<Option>,
+        ): Product {
+            return Product(name = productForm.name, price = productForm.price, imageUrl = productForm.imageUrl)
+        }
```
It is not being used. 🙂

bojana-petroska replied:
yes, I see it now. I will remove it!

Wordbe commented on src/main/kotlin/ecommerce/dto/product/ProductResponse.kt:
```diff
@@ -0,0 +1,8 @@
+package ecommerce.dto.product
+
+data class ProductResponse(
```
.

Wordbe commented on src/main/kotlin/ecommerce/dto/product/TopProductStats.kt:
```diff
@@ -0,0 +1,9 @@
+
+import java.time.LocalDateTime
+
+data class TopProductStats(
```
.

Wordbe commented on src/main/kotlin/ecommerce/model/Order.kt:
```diff
@@ -0,0 +1,51 @@
+    @OneToOne(mappedBy = "order", cascade = [CascadeType.ALL], orphanRemoval = true)
+    var payment: Payment? = null,
+    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
+    val id: Long = 0L,
```
I set the default value to `0L`, but what does `0L` mean? 🤔

bojana-petroska replied:
I used 0L as a default value because id is a non-nullable Long. I could use `val id: Long? = null` but that would mean handling null checks throughout the code

Wordbe commented on src/main/kotlin/ecommerce/model/OrderItem.kt:
```diff
@@ -0,0 +1,29 @@
+    @Column(nullable = false)
+    val quantity: Int,
+    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
+    val id: Long? = 0L,
```
.

Wordbe commented on src/main/kotlin/ecommerce/model/OrderItem.kt:
```diff
@@ -0,0 +1,29 @@
+@Entity
+@Table(name = "order_item")
+class OrderItem(
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "order_id")
+    var order: Order? = null,
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "product_id", nullable = false)
+    val product: Product,
```
It seems like `Order` will never be null, so unlike `Product`, why is this one nullable? 🙂

bojana-petroska replied:
You're right, the order property is logically required for an `OrderItem` and should never be null.
The reason I used nullable here is mainly to satisfy JPA’s requirement for bidirectional relationships..But, maybe there could be a better solution that I didn't manage to find and the tests running smoothly...

Wordbe replied:
What requirements exist for the bidirectional relationship? Making it non-nullable seems safer. 🙂

Wordbe commented on src/main/kotlin/ecommerce/model/Payment.kt:
```diff
@@ -0,0 +1,27 @@
+    @Column(nullable = false)
+    val currency: String,
+    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
+    val id: Long = 0L,
```
.

Wordbe commented on src/main/kotlin/ecommerce/model/Payment.kt:
```diff
@@ -0,0 +1,27 @@
+@Entity
+@Table(name = "payment")
+class Payment(
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "order_id")
+    var order: Order? = null,
```
.

Wordbe commented on src/main/kotlin/ecommerce/model/Payment.kt:
```diff
@@ -0,0 +1,27 @@
+    val paymentIntentId: String,
+    @Column(nullable = false)
+    val amount: Int,
+    @Column(nullable = false)
+    val currency: String,
```
Consider using an enum. 🙂

Wordbe commented on src/main/kotlin/ecommerce/model/Product.kt:
```diff
@@ -25,10 +25,13 @@ class Product(
-        require(options.size == options.map { it.name }.distinct().size) { "Option names must be distinct" }
-        options.forEach { it.product = this }
+    fun addOption(option: Option) {
+        require(option.name.length <= 50) { "Option name must have length up to 50" }
+        require(option.quantity in 1..100_000_000) { "Option quantity must be between 1 and 100,000,000" }
+        require(option.name.matches(Regex("^[a-zA-Z0-9()\\[\\]+\\-&/_ ]+$"))) { "Option name must contain valid characters" }
```
It might be good to consider `name` as a value object. You could do this validation inside it. It's not mandatory, but try it if you have time. 🙂

Wordbe replied:
Maybe later 🙂

Wordbe commented on src/main/kotlin/ecommerce/service/OrderService.kt:
```diff
@@ -0,0 +1,125 @@
+
+        if (option.quantity < req.quantity) {
+            throw IllegalArgumentException("Not enough stock available")
+        }
```
Shall we try making it more concise using the `require` method? 🙂

Wordbe commented on src/main/kotlin/ecommerce/service/OrderService.kt:
```diff
@@ -0,0 +1,125 @@
+
+        if (paymentResponse == null || paymentResponse.id.isBlank()) {
+            throw StripeClientException.from(paymentResponse?.errorMessage ?: "Stripe did not return a valid response.")
+        }
```
How about assigning a role to the `PaymentService`?

Wordbe commented on src/main/kotlin/ecommerce/service/OrderService.kt:
```diff
@@ -0,0 +1,125 @@
+            throw StripeClientException.from(paymentResponse?.errorMessage ?: "Stripe did not return a valid response.")
+        }
+
+        option.quantity -= req.quantity
```
Try expressing it as a behavior within the domain object. 
Also, think about what advantages it might have. 💡

bojana-petroska replied:
Yes, it makes sense! The `Option` should manage its own state and not the service and it can also be reused. Thanks!

Wordbe commented on src/main/kotlin/ecommerce/service/OrderService.kt:
```diff
@@ -0,0 +1,125 @@
+                createdAt = LocalDateTime.now(),
+            )
+
+        val orderItem =
+            OrderItem(
+                product = product,
+                option = option,
+                quantity = req.quantity,
+            )
+        order.addItem(orderItem)
```
I’m curious, can’t it also be created by accepting multiple `orderItem`s when ordering?

bojana-petroska replied:
Yes, in the current state, the order can only be created with only one product and option. I am gonna change it so it can accept multiple `OrderItems`. Thank you!

Wordbe commented on src/main/kotlin/ecommerce/service/OrderService.kt:
```diff
@@ -0,0 +1,125 @@
+        orderRepository.save(order)
+
+        return OrderResponse(
+            success = true,
```
It looks like this field always holds `true`. Is it really necessary? 🧐

bojana-petroska replied:
Yes, that is correct. I actually didn't notice it until now 😊 I believe it is better to catch those errors with global handler, so I will remove the property.

Wordbe commented on src/main/kotlin/ecommerce/service/OrderService.kt:
```diff
@@ -0,0 +1,125 @@
+
+        return OrderResponse(
+            success = true,
+            message = "Order placed successfully",
```
If `OrderResponse` is returned, it means success, so I'm not sure if this message is necessary.

Wordbe commented on src/main/kotlin/ecommerce/service/OrderService.kt:
```diff
@@ -0,0 +1,125 @@
+            OrderResponse(
+                success = true,
+                message = "Order with ${it.items.size} item(s)",
+                paymentIntentId = it.payment?.paymentIntentId ?: "N/A",
```
If not, returning `null` or throwing an exception seems more appropriate. ⚠️

Wordbe commented on src/main/kotlin/ecommerce/service/OrderService.kt:
```diff
@@ -0,0 +1,125 @@
+        }
+    }
+
+    private fun calculateAmount(
+        price: Int,
+        quantity: Int,
+    ): Int {
+        return price * quantity
+    }
```
The `price` in `Product` is a `Double`, so forcibly converting it to `Int` for calculations seems odd. 🤔

bojana-petroska replied:
Yes, correct. I changed the price and consequently the amount as Double.

Wordbe commented on src/main/kotlin/ecommerce/stripe/StripeProperties.kt:
```diff
@@ -0,0 +1,8 @@
+import org.springframework.boot.context.properties.ConfigurationProperties
+
+@ConfigurationProperties("stripe")
+data class StripeProperties(
```
.

Wordbe commented on src/main/resources/application.properties:
```diff
@@ -11,4 +11,6 @@ spring.jpa.hibernate.ddl-auto=create
\ No newline at end of file
+logging.level.org.hibernate.orm.jdbc.bind=TRACE
+
+stripe.secret-key=sk_test_51RshaoQp3kUgTrozcwAE53fg97xuWsVnLKF0Xj8BbcYh3v7RjJaLh2QSih0zCgPhEh1A7hFew4fgz0ryDh4wQJQ2004fomMaco
```
The key is exposed, making it very insecure. Don’t commit it and find a way to hide it. Since the previously issued key has already been leaked, please reissue it. 🔐

bojana-petroska replied:
Yes! Rookie Mistake! 🙈 I issued a new one and used .env for the secret key

Wordbe commented on src/test/kotlin/ecommerce/controller/AuthControllerTest.kt:
```diff
@@ -25,9 +25,11 @@ import org.springframework.transaction.annotation.Transactional
-    @Autowired private val controller: AuthController,
     @Autowired private val memberRepository: MemberRepository,
 ) {
+    @Autowired
+    lateinit var controller: AuthController
```
How about unifying it with the method injected above?

Wordbe reviewed (changes requested):
Hello Bo!

Thank you for carefully writing and submitting your code.
Please check the review and apply the parts you can.
For repeatedly occurring points, I’ve marked them with `.` for your reference.
In particular, it would be great if you could focus on creating `Order` and `OrderServiceTest` since they contain important logic. For the other objects as well, try to strengthen the test code as much as possible.
Thank you! 🙏

bojana-petroska committed:
refactor: apply review feedback

- Externalize Stripe secret using dotenv
- Refactor OrderService to accept multi items
- Update .gitignore and secure environment config
- Added default messages directly in the exception constructors
- Added object conversion with Jackson

Wordbe reviewed (approved):
Hello Bo!

First of all, thank you for completing the code and incorporating the reviews until the end.
The questions I raised during the review were meant for you to reconsider certain points, and it seems you’ve shared your thoughts and applied some of them.
I’ll go ahead and approve this PR and proceed with the merge.
You’ve worked really hard, and I hope the code you wrote and the concepts you learned here will be helpful in your future journey. Great job, and thank you! 🙏
